social/src/controllers/user-controller.ts

```diff
@@ -14,50 +23,104 @@ export class UserController {
   }
 
   static async addTwitterTokens(data: addTwitterTokensData) {
-    const user = await User.findById(data.userID);
+    const user = await User.findById(data.userID).populate('twitter');
     if (!user) {
-      const user = User.build({ _id: data.userID });
-      user.twitter.oauthAccessToken = data.oauthAccessToken;
-      user.twitter.oauthAccessTokenSecret = data.oauthAccessTokenSecret;
-      user.save();
-      return user;
+      const newUser = User.build({ _id: data.userID });
+      const twitterDetails = TwitterController.createTwitterAccountDetails({
+        oauthAccessToken: data.oauthAccessToken,
+        oauthAccessTokenSecret: data.oauthAccessTokenSecret,
+        twitterUserId: data.twitterUserId,
+        twitterScreenName: data.twitterScreenName,
+      });
+      newUser.twitter.push(twitterDetails);
+      newUser.save();
+
+      return newUser;
     } else {
-      return await User.findByIdAndUpdate(
-        data.userID,
-        {
-          twitter: {
-            oauthAccessToken: data.oauthAccessToken,
-            oauthAccessTokenSecret: data.oauthAccessTokenSecret,
-          },
-        },
-        { new: true }
-      );
+      //check if these details are already in db and update if it's the case
+      let exists = false;
+      user.twitter.map(async (twitterAccount) => {
+        if (twitterAccount.twitterUserId === data.twitterUserId) {
+          exists = true;
+          await TwitterController.updateTwitterAccountDetails(
+            twitterAccount.id,
+            {
+              oauthAccessToken: data.oauthAccessToken,
+              oauthAccessTokenSecret: data.oauthAccessTokenSecret,
+              twitterUserId: data.twitterUserId,
+              twitterScreenName: data.twitterScreenName,
+            }
+          );
+        }
+      });
+      if (!exists) {
+        // a new twitter account to connect
+        const twitterDetails = Twitter.build({
+          oauthAccessToken: data.oauthAccessToken,
+          oauthAccessTokenSecret: data.oauthAccessTokenSecret,
+          twitterUserId: data.twitterUserId,
+          twitterScreenName: data.twitterScreenName,
+        });
+        twitterDetails.save();
+        user.twitter.push(twitterDetails);
+        user.save();
+        return user;
+      }
+      return null;
     }
   }
 
-  static async getUserTwitterTokens(userID: string) {
-    const user = await User.findById(userID);
-    if (user) {
-      return {
-        oauthAccessToken: user.twitter.oauthAccessToken,
-        oauthAccessTokenSecret: user.twitter.oauthAccessTokenSecret,
-      };
-    } else {
+  static async getUserAllTwitterAccounts(userId: string) {
+    const user = await User.findById(userId).populate('twitter');
```

Comment:
Refactored `getUserTwitterAccountTokens()`. Now it will return the tokens or throw errors. I had to refactor `getTwitterAccountTokens()` from the service folder as well. Now it only calls `getUserTwitterAccountTokens()` from user controller. Shall I remove this and use the controller directly in routes? Or shall I keep it to avoid importing the controller and calling `getUserTwitterAccountTokens()` inside the routes? What would you suggest?